homeassistant/components/ecobee/config_flow.py

```diff
@@ -0,0 +1,112 @@
+        """Handle a flow initiated by the user."""
+        if self._async_current_entries():
+            # Config entry already exists, only one allowed.
+            return self.async_abort(reason="one_instance_only")
+
+        errors = {}
+        stored_api_key = self.hass.data[DATA_ECOBEE_CONFIG].get(CONF_API_KEY)
+
+        if user_input is not None:
+            # Use the user-supplied API key to attempt to obtain a PIN from ecobee.
+            self._ecobee = Ecobee(config={ECOBEE_API_KEY: user_input[CONF_API_KEY]})
+
+            if await self.hass.async_add_executor_job(self._ecobee.request_pin):
+                # We have a PIN; move to the next step of the flow.
+                return await self.async_step_authorize()
+            errors["base"] = "pin_request_failed"
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {vol.Required(CONF_API_KEY, default=stored_api_key): str}
+            ),
+            errors=errors,
+        )
+
+    async def async_step_authorize(self, user_input=None):
+        """Present the user with the PIN so that the app can be authorized on ecobee.com."""
+        errors = {}
+
+        if user_input is not None:
+            # Attempt to obtain tokens from ecobee and finish the flow.
+            if await self.hass.async_add_executor_job(self._ecobee.request_tokens):
+                # Refresh token obtained; create the config entry.
+                config = {
+                    CONF_API_KEY: self._ecobee.api_key,
+                    CONF_REFRESH_TOKEN: self._ecobee.refresh_token,
+                }
+                return self.async_create_entry(title=DOMAIN, data=config)
+            errors["base"] = "token_request_failed"
+
+        return self.async_show_form(
+            step_id="authorize",
+            errors=errors,
+            description_placeholders={"pin": self._ecobee.pin},
+        )
+
+    async def async_step_import(self, import_data):
+        """
+        Import ecobee config from configuration.yaml.
+
+        Triggered by async_setup only if a config entry doesn't already exist.
+        If ecobee.conf exists, we will attempt to validate the credentials
+        and create an entry if valid. Otherwise, we will delegate to the user
+        step so that the user can continue the config flow.
+        """
+        try:
+            legacy_config = await self.hass.async_add_executor_job(
+                load_json, self.hass.config.path(ECOBEE_CONFIG_FILENAME)
+            )
+            config = {
+                ECOBEE_API_KEY: legacy_config[ECOBEE_API_KEY],
+                ECOBEE_REFRESH_TOKEN: legacy_config[ECOBEE_REFRESH_TOKEN],
+            }
+        except (HomeAssistantError, KeyError):
+            _LOGGER.debug(
+                "No valid ecobee.conf configuration found for import, delegating to user step"
+            )
+            return await self.async_step_user()
```

Comment:
We should pass the api key from config yaml here, right?

Reply:
I’m fine with changing that if you want to,, but my thinking was that we should always show the first step (with the API key filled in on the form that is shown on the user step as is done currently) just in case the user wants to change it without restarting Home Assistant (as they would if the key was hard coded in config.yaml and we skipped showing the form by passing the key in as user_input to async_step_user).

Reply:
If the user uses config yaml they don't want to use the GUI. If they want to use the GUI, the user can remove the config yaml section.

Reply:
Unfortunately they're still going to need to use the GUI since we need to present them the token to authorize on ecobee.com - as they had to do previously via the `configurator` component. I'll make the changes to pass the api key to the user step.